### DISCUSSION TITLE
How to decode an extrinsic in the client?

### DISCUSSION THREAD

**Discussion by nazar-pc** [paritytech/substrate#9823] [Q&A] [answered]
Hey folks, what is the idiomatic way to decode a particular extrinsic out of `OpaqueExtrinsic` in the client?

Currently I have a few nested data structures with indexes matching those generated by proc macros and strategic offset of 3 bytes from the beginning of `OpaqueExtrinsic`'s encoding (that I can't really explain, some of it comes from `Vec<T>`, but the rest - no idea), but it is extremely ugly, there must be a better way.

Maybe I could add some method to runtime API that will be able to take `OpaqueExtrinsic` as input and return `Option<T>` if decoding matches what I want it to be? All examples I found in Substrate repo do the opposite.

**Comment by Voxelot** [comment, marked as answer]
The most straightforward way is to deserialize as an UncheckedExtrinsic that you've configured for your runtime. For example, you could add a `deserialize_opaque_extrinsic` fn to your main  `runtime/lib.rs`. 
```
use parity_scale_codec::{Decode, Encode};
use sp_runtime::generic;

pub type Address = sp_runtime::MultiAddress<AccountId, ()>;
// Where Call comes from `construct_runtime!`
pub type UncheckedExtrinsic = generic::UncheckedExtrinsic<Address, Call, Signature, SignedExtra>;

pub fn deserialize_opaque_extrinsic(opaque_ext: OpaqueExtrinsic) -> Result<UncheckedExtrinsic, parity_scale_codec::Error> {
    let bytes = some_opaque_extrinsic.encode();
    UncheckedExtrinsic::decode(&mut bytes.as_slice())
}
```

Once you have an UncheckedExtrinsic, you can then match on the function to figure out which pallet call it was or inspect the signature to find the public key of the signer etc...

**Reply by nazar-pc** [reply]
Hm, that should work indeed, don't know why I didn't think about it. Thanks!

**Comment by bkchr** [comment]
Decoding an extrinsic from the client is a bad idea, because there could be a runtime upgrade and then you are not being able to decode the extrinsic anymore. The better solution is to use a runtime api to call into the runtime to decode the extrinsic you are interested in and return fixed types.

**Reply by nazar-pc** [reply]
Yes, that is exactly what I ended up doing, thanks!
